File: src/test/java/bad/robot/radiate/monitor/MonitorTest.java
package bad.robot.radiate.monitor;

import org.jmock.Expectations;
import org.jmock.Mockery;
import org.jmock.integration.junit4.JMock;
import org.jmock.integration.junit4.JUnit4Mockery;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.concurrent.RunnableScheduledFuture;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

import static com.googlecode.totallylazy.matchers.IterableMatcher.hasSize;
import static java.util.Arrays.asList;
import static org.hamcrest.MatcherAssert.assertThat;

@RunWith(JMock.class)
public class MonitorTest {

    public final Mockery context = new JUnit4Mockery();

    private final ScheduledExecutorService executor = context.mock(ScheduledExecutorService.class);
    private final MonitoringTask task = context.mock(MonitoringTask.class);
    private final ScheduledFuture future = context.mock(ScheduledFuture.class);
    private final RunnableScheduledFuture waiting = context.mock(RunnableScheduledFuture.class);

    private final Monitor monitor = new ScheduledMonitor(executor);

    @Test
    public void canStopPreviouslyScheduledTasks() {
        context.checking(new Expectations() {{
            allowing(executor).scheduleWithFixedDelay(with(any(Runnable.class)), with(any(Long.class)), with(any(Long.class)), with(any(TimeUnit.class))); will(returnValue(future));
            oneOf(future).cancel(true);
        }});
        Iterable<ScheduledFuture<?>> scheduled = monitor.start(asList(task));
        assertThat(scheduled, hasSize(1));
        monitor.cancel(scheduled);
    }

    @Test
    public void canTerminateMonitoring() {
        context.checking(new Expectations() {{
            oneOf(executor).shutdownNow(); will(returnValue(asList(waiting)));
            oneOf(waiting).cancel(true);
        }});
        monitor.stop();
    }
}

File: src/test/scala/bad/robot/radiate/monitor/MonitorTest.scala
package bad.robot.radiate.monitor

import java.util.Arrays.asList
import java.util.concurrent.{RunnableScheduledFuture, ScheduledExecutorService, ScheduledFuture}

import org.scalamock.specs2.MockContext
import org.specs2.mutable.Specification

class MonitorSTest extends Specification {

  "Can stop previously scheduled tasks" in new MockContext {
    val executor = mock[ScheduledExecutorService]
    val task = mock[MonitoringTaskS]
    val future = mock[ScheduledFuture[Any]]

    (executor.scheduleWithFixedDelay _).expects(*, *, * , *).returning(future)
    (future.cancel _).expects(true).once

    val monitor = new ScheduledMonitorS(executor)

    val scheduled = monitor.start(List(task))
    scheduled must have size 1
    monitor.cancel(scheduled)
  }

  "Can terminate monitoring" in new MockContext {
    val executor = mock[ScheduledExecutorService]
    val task = mock[MonitoringTask]
    val future = mock[ScheduledFuture[Any]]
    val waiting = mock[RunnableScheduledFuture[Any]]

    val monitor = new ScheduledMonitorS(executor)

    (executor.shutdownNow _).expects().returning(asList(waiting)).once
    (waiting.cancel _).expects(true).once

    monitor.stop
  }

}